cmd/maya-apiserver/cstor-operator/cspc/handler.go

```diff
@@ -120,37 +120,45 @@ func (c *Controller) enqueueCSPC(cspc interface{}) {
 }
 
 // synSpc is the function which tries to converge to a desired state for the cspc.
-func (c *Controller) syncCSPC(cspc *apis.CStorPoolCluster) error {
-	//err := validate(cspc)
-	//if err != nil {
-	//	glog.Errorf("Validation of cspc failed:%s", err)
-	//	return nil
-	//}
+func (c *Controller) syncCSPC(cspcGot *apis.CStorPoolCluster) error {
+
 	openebsNameSpace := env.Get(env.OpenEBSNamespace)
 	if openebsNameSpace == "" {
 		message := fmt.Sprint("Could not sync CSPC: got empty namespace for openebs from env variable")
-		c.recorder.Event(cspc, corev1.EventTypeWarning, "Getting Namespace", message)
-		glog.Errorf("Could not sync CSPC {%s}: got empty namespace for openebs from env variable", cspc.Name)
+		c.recorder.Event(cspcGot, corev1.EventTypeWarning, "Getting Namespace", message)
+		glog.Errorf("Could not sync CSPC {%s}: got empty namespace for openebs from env variable", cspcGot.Name)
 		return nil
 	}
 
-	pc, err := c.NewPoolConfig(cspc, openebsNameSpace)
+	pc, err := c.NewPoolConfig(cspcGot, openebsNameSpace)
 	if err != nil {
 		message := fmt.Sprintf("Could not sync CSPC : failed to get pool config: {%s}", err.Error())
-		c.recorder.Event(cspc, corev1.EventTypeWarning, "Creating Pool Config", message)
-		glog.Errorf("Could not sync CSPC {%s}: failed to get pool config: {%s}", cspc.Name, err.Error())
+		c.recorder.Event(cspcGot, corev1.EventTypeWarning, "Creating Pool Config", message)
+		glog.Errorf("Could not sync CSPC {%s}: failed to get pool config: {%s}", cspcGot.Name, err.Error())
 		return nil
 	}
 
-	if !cspc.DeletionTimestamp.IsZero() {
-		// if returns error, we will log the error instead of re-queueing
-		err = c.handleDeletion(cspc, openebsNameSpace)
+	// If CSPC is deleted -- handle the deletion.
+	if !cspcGot.DeletionTimestamp.IsZero() {
+		err = pc.handleCSPCDeletion()
 		if err != nil {
-			glog.Errorf("Could not remove finalizers: %s", err.Error())
+			glog.Errorf("Failed to sync CSPC for deletion:%s", err.Error())
 		}
 		return nil
 	}
 
+	cspcBuilderObj, err := apiscspc.BuilderForAPIObject(cspcGot).Build()
+	if err != nil {
+		glog.Errorf("Failed to build CSPC api object %s", cspcGot.Name)
+		return nil
+	}
+
+	cspc, err := cspcBuilderObj.AddFinalizer(apiscspc.CSPCFinalizer)
```

Comment:
we can push this `cspc` back into `pc`.. and, wherever we calling with `pc`, we can avoid passing `cspc` like in the case of `pc.create(pendingPoolcount, cspc)`